### ISSUE TITLE
BatchSamplerDataCollatorForSeq2Seq does not properly construct batches

### ISSUE THREAD

**Issue by RicardoDominguez** [axolotl-ai-cloud/axolotl#1082]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Batches should be of shape `(micro_batch_size, sequence_len)`

### Current behaviour

Batches are of shape `(1, micro_batch_size*sequence_len)`

### Steps to reproduce

Add `print('Out features shape', features['input_ids'].shape)` in `DataCollatorForSeq2Seq`  before `return features`, add `print("Batch size", len(features)); print("Len of first batch", len(features[0]['input_ids']))` in `BatchSamplerDataCollatorForSeq2Seq` before `return super().__call__(features, return_tensors=return_tensors)`.

Run `accelerate launch -m axolotl.cli.train examples/llama-2/fft_optimized.yml --micro_batch_size 4 --val_set_size 0`

Results in

```
Batch size 1
Len of first batch 16258
Out features shape torch.Size([1, 16384])
```

### Possible solution

If this is indeed not intended behavior I am happy to fix it

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10.12

### axolotl branch-commit

main/d69ba2b

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by RicardoDominguez** [comment]
Note that `DataCollatorForSeq2Seq` does produce `(micro_batch_size, sequence_len)` when sample packing is disabled

**Comment by winglian** [comment]
This is the correct expected behavior. What we are doing is filling as much of a single  match with more data and removing g what would normally be padding tokens

**Comment by RicardoDominguez** [comment]
But given that memory requirements scale quadratically with sequence length, this prevents you from using relatively large batch sizes, right?

Why can't sample packing be applied to the individual batch elements, such that you have `micro_batch_size` elements each with as many documents as one is able to fit?

**Comment by winglian** [comment]
Yeah. I think we could try that as well. It's definitely less efficient to pack data into n smaller context lengths than a single n*context length batch, so that's something to consider. Ultimately we would need to see which is more perfomant for the gpu to process. Is this something you would be interested in tackling?

**Comment by RicardoDominguez** [comment]
Yes, I'll give it a go.

**Comment by RicardoDominguez** [comment]
So after a quick implementation [here](https://github.com/RicardoDominguez/axolotl/tree/sample_packing), it seems that **without flash attention** the current approach is quite suboptimal. As suspected (specially without flash attention), the quadratic cost of increasing context length greatly offsets the benefits of the few percentage points of higher packing efficiency one might get. Further, the training loss is much higher, which is to be expected as it can result in very long sequence lengths

Training Mistral 7B on alpaca_2k_test using a single H100 without flash attention and sequence length 8k, yields the following

| Implementation | Batch size | Grad acc steps | Packing eff | Run time | Loss |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
Current | 4 | 1 | OOM | OOM | OOM |
Current | 2 | 2 | 0.966 | 3:37 | 3.82 | 
Current | 1 |4 | 0.966 | 2:25 | 2.63 |
Proposed | 4 | 1 | 0.932 | 2:25 | 2.70 |

With the current implementation, I would expect that using batch size > 1 is seldom beneficial.

I have so far not been able to get the proposed implementation to work with flash attention, since `get_cu_seqlens_from_pos_ids` gives an error when concatenating. I am quite unfamiliar with the implementation details of flash attention, so if you have any suggestions @winglian I would greatly appreciate it!

**Comment by winglian** [comment]
could you also drop the yaml you used in a GitHub gist and share that please? I think that it would be expected that it won't work without flash attention since the native torch implementations don't know how to handle packing and the cu_seqlens.

**Comment by RicardoDominguez** [comment]
[Here is the config.](https://gist.github.com/RicardoDominguez/d2763778e08eb33e3ff1aafade508e4d) Indeed, perplexity is much smaller with flash attention. I was unaware that packing only works with flash attention, so the comparison above is rather meaningless then.

The bottleneck seems to be adapting flash attention to handle a batch with potentially a different number of samples in each batch element. Do you have any starting points for this?

**Comment by RicardoDominguez** [comment]
When using flash attention, the current implementation seems to be the recommended approach indeed as per https://github.com/Dao-AILab/flash-attention/issues/654

**Comment by winglian** [comment]
@RicardoDominguez btw, It might be worth resurrecting this so that we can use this to enable sample packing for torch's scaled_dot_product_attention, since we can use it with https://huggingface.co/blog/poedator/4d-masks, although the packing efficiency would be slightly less, but still an improvement for most people

**Comment by RicardoDominguez** [comment]
Yes, I think it would be a worthwhile improvement. Do you know if these 4d masks are/will be compatible with flash attention?

**Comment by winglian** [comment]
4d masks won't work with flash-attention since flash attention simply ignores the attention mask.
